### PR TITLE
HIVE-26479: Add ability to set parameters for query-based compaction

### PR DESCRIPTION
**Pull request by SourabhBadhya**
### What changes were proposed in this pull request?

Ability to set parameters for query-based compaction.

### Why are the changes needed?

[HIVE-13354](https://issues.apache.org/jira/browse/HIVE-13354) introduced the ability to set some parameters for the compaction through table properties, like the mapper memory size or compaction thresholds. This could be useful for the query-based compaction as well, for example if the insert of the query-based compaction is failing, we would have a possibility to tune the compaction run directly. First it should be investigated which properties are possible and would make sense to set for the query base compaction. Then implement this feature for the query-based compaction.

### Does this PR introduce _any_ user-facing change?

User can add parameters in TBLPROPERTIES when query-based compaction is configured to run.

### How was this patch tested?

Unit tests

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java [hunk 1]
```diff
@@ -272,5 +277,23 @@ static void removeFilesForMmTable(HiveConf conf, AcidDirectory dir) throws IOExc
         fs.delete(dead, true);
       }
     }
+
+    static void overrideConfProps(HiveConf conf, CompactionInfo ci, Map<String, String> properties) {
+      for (String key : properties.keySet()) {
+        if (key.startsWith(COMPACTOR_PREFIX)) {
+          String property = key.substring(10); // 10 is the length of "compactor." We only keep the rest.
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java:284]: magic number, could we use COMPACTOR_PREFIX.length() instead

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java [hunk 2]
```diff
@@ -272,5 +277,23 @@ static void removeFilesForMmTable(HiveConf conf, AcidDirectory dir) throws IOExc
...
+          conf.set(property, properties.get(key));
+        }
+      }
+
+      // Give preference to properties coming from compaction
+      // over table properties
+      for (String key : ci.getPropertiesMap().keySet()) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java:291]: why is this required? those properties shouldn't affect the execution

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionInfo.java [hunk 3]
```diff
@@ -109,6 +109,13 @@ public void setProperty(String key, String value) {
     properties = propertiesMap.toString();
   }
 
+  public StringableMap getPropertiesMap() {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionInfo.java:112]: not needed

**Review by deniskuzZ — COMMENTED** [review]
LGTM, minor comments

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java (hunk 2 above)
Comment (SourabhBadhya) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java:291]: This is required for the `ALTER TABLE <tablename> COMPACT 'major' WITH TBLPROPERTIES OVERWRITE (<property list>)` scenario. Here the properties are entered into `COMPACTION_QUEUE` and when the table is picked for compaction, the table properties which are set in `CQ_TBLPROPERTIES` must also be picked. All this information is stored in CompactionInfo, hence the need to set the properties as here.

Same behaviour is seen for MR-based compaction as well.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionInfo.java (hunk 3 above)
Comment (SourabhBadhya) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionInfo.java:112]: Please see [this](https://github.com/apache/hive/pull/3528#discussion_r952191010) comment.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java (hunk 1 above)
Comment (SourabhBadhya) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java]: Done.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java (hunk 2 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java]: could we deduplicate this part of the code? something like:
````
static void overrideConfProps(HiveConf conf, CompactionInfo ci, Map<String, String> properties) {
      Stream.of(properties, new StringableMap(ci.properties))
        .filter(Objects::nonNull)
        .flatMap(map -> map.entrySet().stream())
        .filter(entry -> entry.getKey().startsWith(COMPACTOR_PREFIX))
        .forEach(entry -> {
          String property = entry.getKey().substring(COMPACTOR_PREFIX.length());
          conf.set(property, entry.getValue());
        });
    }
````

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionInfo.java (hunk 3 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionInfo.java:112]: see above suggestion :)

Comment (SourabhBadhya) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionInfo.java]: Done.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java (hunk 2 above)
Comment (SourabhBadhya) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java]: Done.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java [hunk 4]
```diff
@@ -272,5 +278,16 @@ static void removeFilesForMmTable(HiveConf conf, AcidDirectory dir) throws IOExc
         fs.delete(dead, true);
       }
     }
+
+    static void overrideConfProps(HiveConf conf, CompactionInfo ci, Map<String, String> properties) {
+      Stream.of(properties, ci.properties != null ? new StringableMap(ci.properties) : null)
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java:283]: no, nee for null check

Comment (SourabhBadhya) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java:283]: There are 2 implementations of StringableMap. One has the null check within, whereas the other one does not have it. I thought there was only one implementation. 
https://github.com/apache/hive/blob/master/common/src/java/org/apache/hadoop/hive/common/StringableMap.java
https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/StringableMap.java

Comment (SourabhBadhya) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java]: Removed the following file - https://github.com/apache/hive/blob/master/common/src/java/org/apache/hadoop/hive/common/StringableMap.java
 
However, regarding the behaviour of toString wherein null is given a length of -1. I have retained the same behaviour moving forward mainly because of [HIVE-17391](https://issues.apache.org/jira/browse/HIVE-17391).

**Review by deniskuzZ — APPROVED** [review]
